### ISSUE TITLE
make GLFW C revision a part of input to Go build cache

### ISSUE THREAD

**Issue by dmitshur** [go-gl/glfw#269]
From https://golang.org/cmd/go/#hdr-Build_and_test_caching:

> The build cache correctly accounts for changes to Go source files, compilers, compiler options, and so on: cleaning the cache explicitly should not be necessary in typical use. However, the build cache does not detect changes to C libraries imported with cgo. If you have made changes to the C libraries on your system, you will need to clean the cache explicitly or else use the -a build flag (see 'go help build') to force rebuilding of packages that depend on the updated C libraries.

There's currently a problem where updating the GLFW C version does not invalidate the Go build cache for the `github.com/go-gl/glfw/v3.3/glfw` package.

We have a `GLFW_C_REVISION.txt` file that tracks the revision of the GLFW C library being used.

We can rename the `GLFW_C_REVISION.txt` file to something like `glfwcrev.go` and include the GLFW C revision as a comment:

```
package glfw

// GLFW_C_REVISION = 0a49ef0a00baa3ab520ddc452f0e3b1e099c5580
```

Then changing the GLFW C revision will cause one of the .go files to become different, and the old build cache will no longer get in the way.

This is somewhat related to #251 and #258.

**Comment by pwaller** [comment]
A friend of mine just hit a related issue, when they tried to diagnose an issue by editing the C files, they found that the build cache meant that it wouldn't actually pick up the new files. My workaround was this: `GODEBUG=gocacheverify=1 go build -v`.

Not ideal. I know we shouldn't be editing the files but it can be useful for diagnostics.

The only other way I'm aware of that we could fix this in principle on the glfw side is to put the C files into the same directory as the package. AIUI, this would result in Go building those C files. So we'd have to have a package per platform, or #ifdef toggles, essentially. I don't see a neat way which allows us to drop in the upstream glfw sources, at a glance.

@dmitshur do you have any other ideas? Do we have to live with the behaviour as it is?

**Comment by dmitshur** [comment]
> @dmitshur do you have any other ideas? Do we have to live with the behaviour as it is?

@pwaller My suggestion is:

> We have a `GLFW_C_REVISION.txt` file that tracks the revision of the GLFW C library being used.
>
>We can rename the `GLFW_C_REVISION.txt` file to something like `glfwcrev.go` and include the GLFW C revision as a comment:
>
> ```Go
> package glfw
>
> // GLFW_C_REVISION = 0a49ef0a00baa3ab520ddc452f0e3b1e099c5580
> ```
>
> Then changing the GLFW C revision will cause one of the .go files to become different, and the old build cache will no longer get in the way.

(I'm not sure if you missed it in the original issue report.)

Maybe it should be an unexported constant instead of a comment. Both are probably fine.

**Comment by dmitshur** [comment]
Oh, I see you were referring to editing C files for debugging specifically.

Given how rarely that happens, and it's done by package developers, I'm okay with them having to be mindful of the build cache for now.

Let's fix this problem for users of the package first (what this issue is about), then we can see if there's a reasonable way to also fix it for developers. I don't think we should do it if it complicates things too much.

**Comment by pwaller** [comment]
I've merged #310 as a solution to this (currently only implemented for v3.3), since #309 upgraded the sources to v3.3.3 and it seems important that infected caches don't confuse things in hard-to-detect ways as users who may have built prior to v3.3.3 try to pull this in.

It addresses both of our use cases, I think, by having the hash reflect the files as they are on disk at the time of invoking go generate. go generate is invoked when `grab-upstream.sh` is run, however, go generate does not yet invoke grab-upstream as we discussed it might (or similar) in [#271](https://github.com/go-gl/glfw/pull/271#issuecomment-593168919).
